## Establish Dependencies and Load in Data

In [1]:
# Import all packages and dependencies
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score
import sys, warnings, os
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

In [2]:
#Ignore warnings when validating scores
if not sys.warnoptions:
    warnings.simplefilter("ignore")
    os.environ["PYTHONWARNINGS"] = "ignore"

In [3]:
#Setting columns and rows to display all the results
pd.set_option("display.max_columns", None, "display.max_rows", None)

In [4]:
# Read in the dataset
ufc_master_df = pd.read_csv("../Resources/ufc-master.csv")
ufc_master_df.head()

R_fighter         B_fighter  R_odds  B_odds        R_ev  \
0   Alistair Overeem  Alexander Volkov     150    -182  150.000000   
1     Cory Sandhagen     Frankie Edgar    -400     300   25.000000   
2  Alexandre Pantoja        Manel Kape    -125     100   80.000000   
3     Diego Ferreira    Beneil Dariush    -125     105   80.000000   
4    Michael Johnson        Clay Guida    -225     175   44.444444   

         B_ev      date                location country Winner  title_bout  \
0   54.945055  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   
1  300.000000  2/6/2021  Las Vegas, Nevada, USA     USA    Red       False   
2  100.000000  2/6/2021  Las Vegas, Nevada, USA     USA    Red       False   
3  105.000000  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   
4  175.000000  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   

   weight_class gender  no_of_rounds  B_current_lose_streak  \
0   Heavyweight   MALE             5                      0   
1  Bantamweight   MALE             3                      0   
2     Flyweight   MALE             3                      0   
3   Lightweight   MALE             3                      0   
4   Lightweight   MALE             3                      2   

   B_current_win_streak  B_draw  B_avg_SIG_STR_landed  B_avg_SIG_STR_pct  \
0                     1       0                  4.76               0.58   
1                     1       1                  3.70               0.39   
2                     0       0                  0.00               0.00   
3                     5       1                  3.98               0.49   
4                     0       0                  2.41               0.33   

   B_avg_SUB_ATT  B_avg_TD_landed  B_avg_TD_pct  B_longest_win_streak  \
0            0.2             0.69          0.70                     4   
1            0.3             2.28          0.31                     5   
2            0.0             0.00          0.00                     0   
3            1.1             1.75          0.32                     5   
4            0.7             3.37          0.38                     4   

   B_losses  B_total_rounds_fought  B_total_title_bouts  \
0         2                     26                    0   
1         8                     94                    9   
2         0                      0                    0   
3         4                     36                    0   
4        15                     87                    0   

   B_win_by_Decision_Majority  B_win_by_Decision_Split  \
0                           0                        1   
1                           0                        1   
2                           0                        0   
3                           0                        1   
4                           0                        3   

   B_win_by_Decision_Unanimous  B_win_by_KO/TKO  B_win_by_Submission  \
0                            2                3                    0   
1                           10                5                    2   
2                            0                0                    0   
3                            4                3                    5   
4                            8                2                    4   

   B_win_by_TKO_Doctor_Stoppage  B_wins  B_Stance  B_Height_cms  B_Reach_cms  \
0                             0       6  Orthodox        200.66       203.20   
1                             0      18  Orthodox        167.64       172.72   
2                             0       0  Southpaw        165.10       172.72   
3                             0      13  Southpaw        177.80       182.88   
4                             0      17  Orthodox        170.18       177.80   

   B_Weight_lbs  R_current_lose_streak  R_current_win_streak  R_draw  \
0           250                      0                     2       0   
1           135                      0                     1       0   
2           

In [5]:
print(ufc_master_df.shape)
print(ufc_master_df.info())

(4566, 137)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4566 entries, 0 to 4565
Columns: 137 entries, R_fighter to B_rev_bout
dtypes: bool(1), float64(73), int64(47), object(16)
memory usage: 4.7+ MB
None


## Baseline Classification Model

In [6]:
# Splitting X (features) from y (labels)
#Classification task, the target value will be "Winner"
y = ufc_master_df.Winner
X = ufc_master_df.drop(['Winner'], axis=1)

In [7]:
# Begin encoding categorical variables by separating variabes by datatype
categorical_col = [col for col in X.columns if X[col].dtypes == 'object']
numeric_col = [col for col in X.columns if col not in categorical_col]

In [8]:
# Encode categorical columns first in order to avoid data leakage in split
enc = LabelEncoder()
for i in X[categorical_col]:
    #using astype(str) to avoid columns with 'float and str' to throw errors
    X[i] = enc.fit_transform(X[i].astype(str))
    

In [9]:
# Encode label (y) for Red fighter ID = 1 Blue fighter ID = 0
y = [1 if win == "Red" else 0 for win in y]


In [10]:
# Split dataset into train test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size = 0.25)

In [11]:
# Find and fill null values
# List column key/values in descending order
X_train.isnull().sum().sort_values(ascending=False)

B_Women's Featherweight_rank    3424
R_Women's Featherweight_rank    3417
B_Pound-for-Pound_rank          3403
B_Women's Flyweight_rank        3396
R_Women's Flyweight_rank        3387
B_Women's Strawweight_rank      3386
B_Featherweight_rank            3365
B_Women's Bantamweight_rank     3365
B_Lightweight_rank              3363
B_Light Heavyweight_rank        3362
B_Bantamweight_rank             3362
B_Flyweight_rank                3358
B_Welterweight_rank             3357
B_Middleweight_rank             3356
R_Women's Strawweight_rank      3356
B_Heavyweight_rank              3346
R_Women's Bantamweight_rank     3343
R_Featherweight_rank            3333
R_Lightweight_rank              3329
R_Middleweight_rank             3329
R_Flyweight_rank                3328
R_Light Heavyweight_rank        3326
R_Bantamweight_rank             3323
R_Welterweight_rank             3317
R_Heavyweight_rank              3317
R_Pound-for-Pound_rank          3312
R_rev_bout                      1330
B

In [12]:
# Creating a StandardScaler instance.

scaler = StandardScaler()
# Fitting the Standard Scaler with the training data.
X_scaler = scaler.fit(X_train)

##Scaling the data.
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [13]:
X_train.sample(10)

R_fighter  B_fighter  R_odds  B_odds        R_ev        B_ev  date  \
4130       1212        138    -295     265   33.898305  265.000000    46   
4126        902       1255     190    -210  190.000000   47.619048    46   
1908       1245        239    -155     135   64.516129  135.000000     4   
622         828        342    -600     450   16.666667  450.000000   389   
4136        634        816    -355     320   28.169014  320.000000    55   
2859       1016       1056   -1000     800   10.000000  800.000000   118   
4153       1123       1360    -174     164   57.471264  164.000000    28   
2184        516        823    -160     140   62.500000  140.000000   330   
3426        928       1305    -145     135   68.965517  135.000000    90   
1804        140       1185     270    -330  270.000000   30.303030   207   

      location  country  title_bout  weight_class  gender  no_of_rounds  \
4130        66       21       False             2       1             3   
4126        66       21       False             4       1             3   
1908       100       21       False             4       1             3   
622         35        8       False             6       1             3   
4136        58       21        True             2       1             5   
2859        13        2       False             0       1             3   
4153       141       21       False             6       1             3   
2184        66       21       False             6       1             3   
3426        45       21       False             8       1             3   
1804        63       21       False             6       1             3   

      B_current_lose_streak  B_current_win_streak  B_draw  \
4130                      0                     0       0   
4126                      2                     0       0   
1908                      1                     0       0   
622                       1                     0       0   
4136                      0                     1       0   
2859                      0                     4       0   
4153                      1                     0       0   
2184                      0                     1       0   
3426                      2                     0       0   
1804                      1                     0       0   

      B_avg_SIG_STR_landed  B_avg_SIG_STR_pct  B_avg_SUB_ATT  B_avg_TD_landed  \
4130                   NaN                NaN            NaN              NaN   
4126             19.500000           0.405000       0.000000         0.250000   
1908             59.000000           0.350000       0.000000         0.000000   
622               1.520000           0.610000       0.000000         3.500000   
4136             25.875000           0.452500       0.812500         1.562500   
2859             22.200000           0.418000       1.600000         1.000000   
4153             29.000000           0.513333       0.166667         1.500000   
2184             43.857143           0.358571       1.000000         2.142857   
3426             23.857143           0.331429       0.714286         0.285714   
1804             64.200000           0.530000       0.200000         0.800000   

      B_avg_TD_pct  B_longest_win_streak  B_losses  B_total_rounds_fought  \
4130           NaN                     0         0                      0   
4126      0.125000                     2         2                      8   
1908      0.000000                     0         1                      3   
622       1.000000                     0         1                      2   
4136      0.445000                     6         4                     37   
2859      0.372000                     4         1                      9   
4153      0.323333                     1         3                     16   
2184      0.348571                     4         2                     17   
3426      0.057143                     4         3                     15   
1804     

In [14]:
# Dummy classifier used for simple model for comparison after optimization-DummyClassifier will allow us to run model with null values
simple_model = DummyClassifier(random_state=42)
simple_model.fit(X_train, y_train)

DummyClassifier(random_state=42)

In [15]:
prediction = simple_model.predict(X_test)
accuracy_score(y_test, prediction)

0.5175131348511384

#### Our simple model can predict winners with ~ 51.75% accuracy

In [16]:
fillna = ufc_master_df.fillna(0)

In [17]:
fillna.reset_index(level=0, inplace=True)
fillna['level_0'] = fillna['index']
fillna.head()

index          R_fighter         B_fighter  R_odds  B_odds        R_ev  \
0      0   Alistair Overeem  Alexander Volkov     150    -182  150.000000   
1      1     Cory Sandhagen     Frankie Edgar    -400     300   25.000000   
2      2  Alexandre Pantoja        Manel Kape    -125     100   80.000000   
3      3     Diego Ferreira    Beneil Dariush    -125     105   80.000000   
4      4    Michael Johnson        Clay Guida    -225     175   44.444444   

         B_ev      date                location country Winner  title_bout  \
0   54.945055  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   
1  300.000000  2/6/2021  Las Vegas, Nevada, USA     USA    Red       False   
2  100.000000  2/6/2021  Las Vegas, Nevada, USA     USA    Red       False   
3  105.000000  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   
4  175.000000  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   

   weight_class gender  no_of_rounds  B_current_lose_streak  \
0   Heavyweight   MALE             5                      0   
1  Bantamweight   MALE             3                      0   
2     Flyweight   MALE             3                      0   
3   Lightweight   MALE             3                      0   
4   Lightweight   MALE             3                      2   

   B_current_win_streak  B_draw  B_avg_SIG_STR_landed  B_avg_SIG_STR_pct  \
0                     1       0                  4.76               0.58   
1                     1       1                  3.70               0.39   
2                     0       0                  0.00               0.00   
3                     5       1                  3.98               0.49   
4                     0       0                  2.41               0.33   

   B_avg_SUB_ATT  B_avg_TD_landed  B_avg_TD_pct  B_longest_win_streak  \
0            0.2             0.69          0.70                     4   
1            0.3             2.28          0.31                     5   
2            0.0             0.00          0.00                     0   
3            1.1             1.75          0.32                     5   
4            0.7             3.37          0.38                     4   

   B_losses  B_total_rounds_fought  B_total_title_bouts  \
0         2                     26                    0   
1         8                     94                    9   
2         0                      0                    0   
3         4                     36                    0   
4        15                     87                    0   

   B_win_by_Decision_Majority  B_win_by_Decision_Split  \
0                           0                        1   
1                           0                        1   
2                           0                        0   
3                           0                        1   
4                           0                        3   

   B_win_by_Decision_Unanimous  B_win_by_KO/TKO  B_win_by_Submission  \
0                            2                3                    0   
1                           10                5                    2   
2                            0                0                    0   
3                            4                3                    5   
4                            8                2                    4   

   B_win_by_TKO_Doctor_Stoppage  B_wins  B_Stance  B_Height_cms  B_Reach_cms  \
0                             0       6  Orthodox        200.66       203.20   
1                             0      18  Orthodox        167.64       172.72   
2                             0       0  Southpaw        165.10       172.72   
3                             0      13  Southpaw        177.80       182.88   
4                             0      17  Orthodox        170.18       177.80   

   B_Weight_lbs  R_current_lose_streak  R_current_win_streak  R_draw  \
0           250                      0                     2       0   
1           135                   

In [19]:
# Saving csv with null values filled
file_path = "../Resources/na_filled.csv"
fillna.to_csv(file_path, index=False, header=True)